plugins/glustershd/rest.go

```diff
@@ -1,197 +1,122 @@
+func selfhealInfoHandler(w http.ResponseWriter, r *http.Request) {
 	p := mux.Vars(r)
 	volname := p["name"]
+	option := p["opts"]
+	_ = option
 
 	ctx := r.Context()
 	logger := gdctx.GetReqLogger(ctx)
 
-	//validate volume name
-	v, err := volume.GetVolume(volname)
-	if err != nil {
-		if err == errors.ErrVolNotFound {
-			restutils.SendHTTPError(ctx, w, http.StatusNotFound, errors.ErrVolNotFound)
+	lock, unlock := transaction.CreateLockFuncs(volname)
+	// Taking a lock outside the txn as volinfo.Nodes() must also
+	// be populated holding the lock.
+	if err := lock(ctx); err != nil {
+		if err == transaction.ErrLockTimeout {
+			restutils.SendHTTPError(ctx, w, http.StatusConflict, err)
 		} else {
 			restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err)
 		}
 		return
 	}
-	// Store initial volinfo before changing the HealFlag
-	tmp := *v
-	oldvolinfo := &tmp
+	defer unlock(ctx)
 
-	// validate volume type
-	if !isVolReplicate(v.Type) {
-		restutils.SendHTTPError(ctx, w, http.StatusBadRequest, "Volume Type not supported")
-		return
+        // Validate volume existence
+	volinfo, err := volume.GetVolume(volname)
+	if err != nil {
+		if err == gderrors.ErrVolNotFound {
+			logger.WithError(err).WithField(
+				"volname", volname).Debug("volume not found")
+                        restutils.SendHTTPError(ctx, w, http.StatusNotFound, err)
+		} else {
+			logger.WithError(err).WithField(
+				"volname", "volname").Debug("error occured while looking for volume")
+                        restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err)
+		}
 	}
 
-	if v.State != volume.VolStarted {
-		restutils.SendHTTPError(ctx, w, http.StatusBadRequest, "Volume should be in started state.")
-		return
-
+        // Validate volume type
+	if !afr.IsVolReplicate(volinfo.Type) {
+		logger.WithError(err).WithField(
+			"volname", volname).Debug("volume type is not supported")
+		 restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err)
 	}
 
-	// Transaction which starts self heal daemon on all nodes with atleast one brick.
-	txn := transaction.NewTxn(ctx)
-	defer txn.Cleanup()
-
-	//Lock on Volume Name
-	lock, unlock, err := transaction.CreateLockSteps(volname)
-	if err != nil {
+        // Validate volume state
+	if volinfo.State != volume.VolStarted {
+		logger.WithError(err).WithField(
+			"volname", volname).Debug("volume should be in started state")
 		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err)
-		return
 	}
 
-	v.HealEnabled = true
+	txn := transaction.NewTxn(ctx)
+	defer txn.Cleanup()
 
-	txn.Nodes = v.Nodes()
 	txn.Steps = []*transaction.Step{
-		lock,
-		{
-			DoFunc:   "vol-option.UpdateVolinfo",
-			Nodes:    []uuid.UUID{gdctx.MyUUID},
-			UndoFunc: "selfheald-undo",
-		},
-		{
-			DoFunc: "selfheal-start",
-			Nodes:  txn.Nodes,
-		},
 		{
-			DoFunc: "vol-option.NotifyVolfileChange",
-			Nodes:  txn.Nodes,
+			DoFunc: "selfheal-info",
+			Nodes:  []uuid.UUID{gdctx.MyUUID},
```

Comment:
Why is this run on just this node ?

If it's indeed supposed to run only on this node, you don't need a transaction.